tests/listobjects/listobjects.go

```diff
@@ -91,128 +99,152 @@ func runTests(t *testing.T, schemaVersion string, client ClientInterface) {
 	err = yaml.Unmarshal(b, &testCases)
 	require.NoError(t, err)
 
-	ctx := context.Background()
 	for _, test := range testCases.Tests {
 		test := test
+		runTest(t, test, testInformation, false)
+		runTest(t, test, testInformation, true)
 
-		t.Run(test.Name, func(t *testing.T) {
-			t.Parallel()
-			resp, err := client.CreateStore(ctx, &pb.CreateStoreRequest{Name: test.Name})
-			require.NoError(t, err)
+	}
+}
 
-			storeID := resp.GetId()
+func runTest(t *testing.T, test individualTest, testInformation testInformation, contextTupleTest bool) {
+	schemaVersion := testInformation.schemaVersion
+	client := testInformation.client
+	ctx := context.Background()
+	name := test.Name
+	if contextTupleTest {
+		name += "_ctxTuples"
+	}
 
-			for _, stage := range test.Stages {
-				// arrange: write model
-				var typedefs []*pb.TypeDefinition
-				if schemaVersion == typesystem.SchemaVersion1_1 {
-					typedefs = parser.MustParse(stage.Model)
+	if contextTupleTest && len(test.Stages) > 1 {
+		// we don't want to run special contextual tuples test for these cases
+		// as multi-stages test has expectation tuples are in system
+		return
+	}
 
-				} else {
-					typedefs = v1parser.MustParse(stage.Model)
-				}
+	t.Run(name, func(t *testing.T) {
+		t.Parallel()
+		resp, err := client.CreateStore(ctx, &pb.CreateStoreRequest{Name: name})
+		require.NoError(t, err)
+
+		storeID := resp.GetId()
+
+		for _, stage := range test.Stages {
+			// arrange: write model
+			var typedefs []*pb.TypeDefinition
+			if schemaVersion == typesystem.SchemaVersion1_1 {
+				typedefs = parser.MustParse(stage.Model)
 
-				_, err = client.WriteAuthorizationModel(ctx, &pb.WriteAuthorizationModelRequest{
-					StoreId:         storeID,
-					SchemaVersion:   schemaVersion,
-					TypeDefinitions: typedefs,
+			} else {
+				typedefs = v1parser.MustParse(stage.Model)
+			}
+
+			_, err = client.WriteAuthorizationModel(ctx, &pb.WriteAuthorizationModelRequest{
+				StoreId:         storeID,
+				SchemaVersion:   schemaVersion,
+				TypeDefinitions: typedefs,
+			})
+			require.NoError(t, err)
+
+			// arrange: write tuples
+			if len(stage.Tuples) > 0 && !contextTupleTest {
+				_, err = client.Write(ctx, &pb.WriteRequest{
+					StoreId: storeID,
+					Writes:  &pb.TupleKeys{TupleKeys: stage.Tuples},
 				})
 				require.NoError(t, err)
+			}
 
-				// arrange: write tuples
-				if len(stage.Tuples) > 0 {
-					_, err = client.Write(ctx, &pb.WriteRequest{
-						StoreId: storeID,
-						Writes:  &pb.TupleKeys{TupleKeys: stage.Tuples},
-					})
-					require.NoError(t, err)
-				}
+			for _, assertion := range stage.ListObjectAssertions {
+				// assert 1: on regular list objects endpoint
+				detailedInfo := fmt.Sprintf("ListObject request: %s. Contextual tuples: %s", assertion.Request, assertion.ContextualTuples)
```

Comment:
```suggestion
				
				detailedInfo := fmt.Sprintf("ListObject request: %s. Contextual tuples: %s", assertion.Request, assertion.ContextualTuples)
				ctxTuples := assertion.ContextualTuples
				if contextTupleTest {
					ctxTuples = append(ctxTuples, stage.Tuples...)
				}
				
				// assert 1: on regular list objects endpoint
```

Reply:
(move the comment down because the code before is purely setup and not doing any assertions)